hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/utils/AutoCompletion.java

```diff
@@ -0,0 +1,164 @@
+import org.apache.hadoop.ozone.s3.Gateway;
+import org.apache.hadoop.ozone.shell.OzoneRatis;
+import org.apache.hadoop.ozone.shell.OzoneShell;
+import org.apache.hadoop.ozone.shell.checknative.CheckNative;
+import org.apache.hadoop.ozone.shell.s3.S3Shell;
+import org.apache.hadoop.ozone.shell.tenant.TenantShell;
+import picocli.AutoComplete;
+import picocli.CommandLine;
+import picocli.CommandLine.Option;
+
+/**
+ * Tool to generate bash/zsh auto-complete scripts for Ozone CLI.
+ */
+public final class AutoCompletion {
+
+  private AutoCompletion() { }
+
+  public static void main(String[] args) {
+
+    CommandLine hierarchy = new CommandLine(new Ozone())
+        .addSubcommand("admin", new OzoneAdmin().getCmd())
+        .addSubcommand("auditparser", new AuditParser().getCmd())
+        .addSubcommand("checknative", new CheckNative().getCmd())
+        .addSubcommand("csi", new CsiServer().getCmd())
+        .addSubcommand("datanode", new HddsDatanodeService().getCmd())
+        .addSubcommand("debug", new OzoneDebug().getCmd())
+        .addSubcommand("freon", new Freon().getCmd())
+        .addSubcommand("genconf", new GenerateOzoneRequiredConfigurations().getCmd())
+        .addSubcommand("getconf", new OzoneGetConf().getCmd())
+        .addSubcommand("httpfs", new HttpFSServerWebServer())
+        .addSubcommand("insight", new Insight().getCmd())
+        .addSubcommand("om", new OzoneManagerStarter(null).getCmd())
+        .addSubcommand("ratis", new OzoneRatis().getCmd())
+        .addSubcommand("recon", new ReconServer().getCmd())
+        .addSubcommand("repair", new OzoneRepair().getCmd())
+        .addSubcommand("s3", new S3Shell().getCmd())
+        .addSubcommand("s3g", new Gateway().getCmd())
+        .addSubcommand("scm", new StorageContainerManagerStarter(null).getCmd())
+        .addSubcommand("sh", new OzoneShell().getCmd(), "shell")
+        .addSubcommand("tenant", new TenantShell().getCmd())
+        .addSubcommand("version", new OzoneVersionInfo());
+
+    /*
+     * The following commands are marked as hidden,
+     * so they won't have auto complete.
+     *   - OzoneAdmin
+     *   - Insight
+     *   - CsiServer
+     *   - HddsDatanodeService
+     *   - OzoneManagerStarter
+     *   - Gateway
+     *   - StorageContainerManagerStarter
+     */
+
+    /*
+     * We don't have auto-complete support for the following commands
+     *   - dtutil    -- hadoop command
+     *   - classpath -- handled in shell script, we can introduce a dummy class here
+     *   - envvars   -- handled in shell script, we can introduce a dummy class here
+     *   - fs        -- hadoop command
+     *   - daemonlog -- hadoop command
+     */
+
+    System.out.println(AutoComplete.bash("ozone", hierarchy));
+  }
+
+  /**
+   * Ozone top level command, used only to generate auto-complete.
+   */
+  @CommandLine.Command(name = "ozone",
+          description = "Ozone top level command")
+  private static class Ozone {
```

Comment:
Also in follow-up, I think we can convert this to a real command and reduce duplication in `ozone` shell script by delegating more tasks (like usage help generation) to this new command.

Reply:
It could also be used to generate manpages or manpage-like HTML doc.

https://issues.apache.org/jira/browse/HDDS-12522